managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ReadOnlyManagedLedgerImpl.java

```diff
@@ -0,0 +1,127 @@
+import org.apache.bookkeeper.mledger.ManagedLedgerException.MetaStoreException;
+import org.apache.bookkeeper.mledger.ManagedLedgerException.MetadataNotFoundException;
+import org.apache.bookkeeper.mledger.ReadOnlyCursor;
+import org.apache.bookkeeper.mledger.impl.MetaStore.MetaStoreCallback;
+import org.apache.bookkeeper.mledger.impl.MetaStore.Stat;
+import org.apache.bookkeeper.mledger.proto.MLDataFormats.ManagedLedgerInfo;
+import org.apache.bookkeeper.mledger.proto.MLDataFormats.ManagedLedgerInfo.LedgerInfo;
+
+public class ReadOnlyManagedLedgerImpl extends ManagedLedgerImpl {
+
+    public ReadOnlyManagedLedgerImpl(ManagedLedgerFactoryImpl factory, BookKeeper bookKeeper, MetaStore store,
+            ManagedLedgerConfig config, OrderedScheduler scheduledExecutor, OrderedExecutor orderedExecutor,
+            String name) {
+        super(factory, bookKeeper, store, config, scheduledExecutor, orderedExecutor, name);
+    }
+
+    CompletableFuture<ReadOnlyCursor> initializeAndCreateCursor(PositionImpl startPosition) {
+        CompletableFuture<ReadOnlyCursor> future = new CompletableFuture<>();
+
+        // Fetch the list of existing ledgers in the managed ledger
+        store.getManagedLedgerInfo(name, false, new MetaStoreCallback<ManagedLedgerInfo>() {
+            @Override
+            public void operationComplete(ManagedLedgerInfo mlInfo, Stat stat) {
+                state = State.LedgerOpened;
+
+                for (LedgerInfo ls : mlInfo.getLedgerInfoList()) {
+                    ledgers.put(ls.getLedgerId(), ls);
+                }
+
+                // Last ledger stat may be zeroed, we must update it
+                if (ledgers.size() > 0 && ledgers.lastEntry().getValue().getEntries() == 0) {
+                    long lastLedgerId = ledgers.lastKey();
+
+                    // Fetch last add confirmed for last ledger
+                    bookKeeper.newOpenLedgerOp().withRecovery(false).withLedgerId(lastLedgerId)
+                            .withDigestType(config.getDigestType()).withPassword(config.getPassword()).execute()
+                            .thenAccept(readHandle -> {
+                                readHandle.readLastAddConfirmedAsync().thenAccept(lastAddConfirmed -> {
+                                    LedgerInfo info = LedgerInfo.newBuilder().setLedgerId(lastLedgerId)
+                                            .setEntries(lastAddConfirmed + 1).setSize(readHandle.getLength())
+                                            .setTimestamp(clock.millis()).build();
+                                    ledgers.put(lastLedgerId, info);
+
+                                    future.complete(createReadOnlyCursor(startPosition));
+                                }).exceptionally(ex -> {
+                                    if (ex instanceof CompletionException
+                                            && ex.getCause() instanceof IllegalArgumentException) {
```

Comment:
Does readLastAddConfirmed throw IllegalArgumentException in real life? or just the mock?

Reply:
The real BK client throws `IllegalArgumentException` and the mock throws `ArrayIndexOutOfBoundException`. Ideally `asyncReadLastAddConfirmed()` should return a `-1` when the ledger is empty